sdk/identity/azure-identity/test/ut/chained_token_credential_test.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/identity/chained_token_credential.hpp"
+
+#include <string>
+
+#include <gtest/gtest.h>
+
+using Azure::Identity::ChainedTokenCredential;
+
+using Azure::Core::Context;
+using Azure::Core::Credentials::AccessToken;
+using Azure::Core::Credentials::AuthenticationException;
+using Azure::Core::Credentials::TokenCredential;
+using Azure::Core::Credentials::TokenRequestContext;
+
+namespace {
+class TestCredential : public TokenCredential {
+private:
+  std::string m_token;
+
+public:
+  TestCredential(std::string token = "") : m_token(token) {}
+
+  mutable bool WasInvoked = false;
```

Comment:
nit question: why is this a mutable bool?

Reply:
Because we do modify it from the `GetToken() const`, i.e. from the const function. Without mutable, the compiler won't allow you to modify the class data member from a const member function.
Other option would be `const_cast`, but `mutable` is better than const_cast - it is designed for exactly this case, plus theoretically complier may not always see all the const_casts, and it could think it is possible to put the object in a readonly memory. And then if your code tries to modify a variable that is in read-only memory, it is UB.
With mutable, compiler will know to not pace the variable to read-only memory.

Reply:
Is changing a function from `const` to `non-const` non-breaking?

Since this is a test and not product code, this coding pattern is fine, but we had a conversation previously about conveying intent to the caller and about use of mutable. If a method is marked as `const`, we should honor that to match caller expectations, and not have mutable fields that are changed underneath. When such a scenario exists in public API of the SDK itself, we should not mark it as const.